File: rust/src/problems/p55.rs
pub struct Solution;

impl Solution {
    pub fn can_jump(nums: Vec<i32>) -> bool {
        let n = nums.len();
        let mut reach = 1;
        for (i, a) in nums.iter().enumerate() {
            if reach <= i {
                return false;
            }
            reach = reach.max(i + 1 + *a as usize);
            if reach >= n {
                return true;
            }
        }
        false
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn case1() {
        let nums = vec![2, 3, 1, 1, 4];
        assert!(Solution::can_jump(nums));
    }

    #[test]
    fn case2() {
        let nums = vec![3, 2, 1, 0, 4];
        assert!(!Solution::can_jump(nums));
    }
}

File: cpp/p55.cpp
#include <algorithm>
#include <cassert>
#include <vector>

using namespace std;

class Solution {
public:
    bool canJump(vector<int> &nums) {
        int f = 0, n = nums.size();
        for (int i = 0; i < n; i++) {
            if (i > f) {
                return false;
            }
            f = max(f, i + nums[i]);
            if (f >= n - 1) {
                return true;
            }
        }
        return false;
    }
};

int main() {
    vector<int> nums = {2, 3, 1, 1, 4};
    assert(Solution().canJump(nums));
    nums = {3, 2, 1, 0, 4};
    assert(!Solution().canJump(nums));
    return 0;
}
